tools/syz-declextract/syz-declextract.cpp

```diff
@@ -64,6 +106,230 @@ bool beginsWith(const std::string_view &str, const std::string_view begin) {
+      return getFieldType(llvm::dyn_cast<CountAttributedType>(field)->desugar(), context, fieldName);
+    case clang::Type::BTFTagAttributed: // Currently Unused
+      return getFieldType(llvm::dyn_cast<BTFTagAttributedType>(field)->desugar(), context, fieldName);
+    case clang::Type::Elaborated:
+      return getFieldType(llvm::dyn_cast<ElaboratedType>(field)->desugar(), context, fieldName, parent, isSyscallParam);
+    case clang::Type::Enum: {
+      const auto &enumDecl = llvm::dyn_cast<EnumType>(field)->getDecl();
+      auto name = enumDecl->getNameAsString();
+      flags.push_back(name);
+      includes.push_back(std::filesystem::relative(SM->getFilename(enumDecl->getSourceRange().getBegin()).str()));
+      const char *sep = " = ";
+      for (const auto &enumerator : enumDecl->enumerators()) {
+        flags.back() += sep + enumerator->getNameAsString();
+        sep = ", ";
+      }
+      return "flags[" + name + "]";
+    }
+    case clang::Type::FunctionProto:
+      return makePtr("in", "int8");
+    default:
+      field->dump();
+      fprintf(stderr, "Unhandled field type %s\n", field->getTypeClassName());
+      exit(1);
+    }
+  }
+
+  std::string extractRecord(const RecordDecl *recordDecl, ASTContext *context, const std::string &backupName = "") {
+    recordDecl = recordDecl->getDefinition();
+    if (!recordDecl) { // When the definition is in a different translation unit.
+      return "intptr";
+    }
+    const auto &name = (recordDecl->getNameAsString().empty() ? backupName : recordDecl->getNameAsString());
+    const auto &recordName = name + "$auto_record";
+    if (visitedRecord.find(name) != visitedRecord.end()) { // Don't extract the same record twice.
+      return recordName;
+    }
+    visitedRecord[name] = extractedRecords.size();
+    extractedRecords.resize(extractedRecords.size() + 1);
+    bool isVarlen = false;
+    std::vector<StructMember> members;
+    for (const auto &field : recordDecl->fields()) {
+      std::string fieldName;
+      if (field->getName().empty()) {
+        fieldName = name + "_" + std::to_string(field->getFieldIndex());
+      } else if (field->isAnonymousStructOrUnion()) {
+        fieldName = name;
+      } else {
+        fieldName = field->getNameAsString();
+      }
+      const std::string &parentName = field->isAnonymousStructOrUnion() ? "" : name;
+      std::string fieldType = getFieldType(field->getType(), context, fieldName, parentName);
+      if (field->isUnnamedBitField()) { // pad
+        fieldType = makeConst(fieldType);
+      } else if (field->isBitField()) {
+        fieldType = fieldType + ":" + std::to_string(field->getBitWidthValue(*context));
+      }
+      if (fieldType.empty()) {
+        continue;
+      }
+      isVarlen |= isFieldVarlen(field->getType()) || (visitedRecord.find(fieldName) != visitedRecord.end() &&
+                                                      extractedRecords[visitedRecord[fieldName]].isVarlen);
```

Comment:
Is it guaranteed that we already processed the fieldName record and it's in extractedRecords? Can it be the case that it is varlen, but we have not processed it yet, so we don't know yet?

Reply:
If the field is a struct, it should be processed and extracted in the fieldType = getFieldType() assignment above this, before we try to check.
getFieldType() calls extractRecord(), and inserts the record.